### PR TITLE
Feature: Tekton task "push-application-snapshot"

### PR DESCRIPTION
**Pull request by parasense**
jira: HACBS-564

Signed-off-by: Jon Disnard <jdisnard@redhat.com>

### CODE REVIEW

**Comment by parasense** [comment]
@davidmogar @johnbieren PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: push-application-snapshot-task
+spec:
+  params:
+    - name: mappedApplicationSnapshot
+      description: |
+        JSON string representing the ApplicationSnapshot
+        For example:
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:11]: Too much info here. Could you take a look to #25? We are adding a README.md file with a short description. Copy the one that Johnny has there and change what's needed please. Easier for external users to see how this works.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        Note:
+        * The pullspec digest (e.g. @sha256:...) may be ommited.
+        * If the pullspec digest is ommited, then assume "latest".
+      type: string
+    - name: retries
+      description: Retry copy N times.
+      type: int
+      default: 0
+    - name: destinationTag
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:37]: This comes in the json already. It's called `tag`.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      description: option tag to give destination images.
+      type: string
+      default: ""
+  steps:
+    - name: mappedApplicationSnapshot
+      image: http://quay.io/hacbs-release/release-utils:latest
+      script: |
+        if [[ -z "$(params.mappedApplicationSnapshot)" ]]
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:45]: This can't happen. That param is not optional in this task. You can safely remove it.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        then
+          exit 1 # Missing URL parameters
+        fi
+        app=$(jq -r '.application' <<<$(params.mappedApplicationSnapshot))
+        tag="$(params.destinationTag)"
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:50]: See above.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        printf 'Begining push task for application: "%s"\n' "$app"
+        while read line;
+        do
+          set -- $(jq -cr '.[]' <<<"$line" | xargs)
+          s_Digest=$(skopeo inspect --format '{{.Digest}}' "$2" 2>/dev/null)
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:55]:
```suggestion
          source_digest=$(skopeo inspect --format '{{.Digest}}' "$2" 2>/dev/null)
```

Note: I didn't know about `--format`. That's nice!

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          d_Digest=$(skopeo inspect --format '{{.Digest}}' "$3" 2>/dev/null)
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:56]:
```suggestion
          destination_digest=$(skopeo inspect --format '{{.Digest}}' "$3" 2>/dev/null)
```

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 3 above)
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:45]: Please, add also a shebang and set flags to exit on error.

Let's go for something like:
```
#!/usr/bin/env sh
set -eu
```

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        if [[ -z "$(params.mappedApplicationSnapshot)" ]]
+        then
+          exit 1 # Missing URL parameters
+        fi
+        app=$(jq -r '.application' <<<$(params.mappedApplicationSnapshot))
+        tag="$(params.destinationTag)"
+        printf 'Begining push task for application: "%s"\n' "$app"
+        while read line;
+        do
+          set -- $(jq -cr '.[]' <<<"$line" | xargs)
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:54]: If I got it correctly, you are setting here new params and expecting `$1` to be the `pullspec` and `$2` to be the `repository`. Thing is that json have no order, so this is a bit dangerous. I prefer you to set variables by quering the json itself.

```
pullspec=$(jq -rc '.pullspec' <<< "$line" | xargs)
repository=$(jq -rc '.repository' <<< "$line" | xargs)
```
I get that you want to be agnostic to field name changes, but given the order issue I think we have to go for this.

**Review by davidmogar — CHANGES_REQUESTED** [review]

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 3 above)
Comment (parasense) [reply definitions/push-application-snapshot/push-application-snapshot.yaml]: fixed - PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 4 above)
Comment (parasense) [reply definitions/push-application-snapshot/push-application-snapshot.yaml]: fixed - PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 3 above)
Comment (parasense) [reply definitions/push-application-snapshot/push-application-snapshot.yaml]: removed - PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 2 above)
Comment (parasense) [reply definitions/push-application-snapshot/push-application-snapshot.yaml]: fixed, tag is now sourced from json. - PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml (hunk 1 above)
Comment (parasense) [reply definitions/push-application-snapshot/push-application-snapshot.yaml]: agreed - didn't know we could put markdown in bundle dirs, and I really like that!
Fully documented - PTAL

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    - name: mappedApplicationSnapshot
+      description: JSON string representing the ApplicationSnapshot
+      type: string
+    - name: retries
+      description: Retry copy N times.
+      type: int
+      default: 0
+  steps:
+    - name: skopeo-copy
+      onError: stopAndFail
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:17]: This is the default behavior isn't it? You can remove it.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+              --all \
+              --preserve-digests \
+              --dest-precompute-digests \
+              --retry-times="$(params.retries)" \
+              "docker://${map[pullspec]}" \
+              "docker://${map[repository]}${map[tag]:+:${map[tag]}}" ;
+          else
+            printf '* Skipping component: %s\n' "${map[component]}"
+          fi
+          unset map
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:47]: I wouldn't unset it. This is the end of the script and the pod is going to go away. Also, with the current code, any failure in a skopeo task will prevent this line to be executed. If you really want to run it, you should use a trap.

File: definitions/push-application-snapshot/push-application-snapshot.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: push-application-snapshot-task
```
Comment (davidmogar) [inline definitions/push-application-snapshot/push-application-snapshot.yaml:5]: We are not using the `-task` suffix.
```suggestion
  name: push-application-snapshot
```

**Review by davidmogar — CHANGES_REQUESTED** [review]
I haven't actually tested the task itself yet but it looks on point. I'll give it a try during the weekend.

**Comment by davidmogar** [comment]
I left a comment for you on Slack. Not sure you have seen it. I sent there an extensive review after testing this in my local cluster.

**Comment by parasense** [comment]
Thanks, I'll check it out within the next hour

File: definitions/push-application-snapshot/README.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    }
+  ]
+}
+```
+| Key | Value | type |
+|------|-------------|------|
+| application | The name of the application | string |
+| images | List of image mappings | array |
+| *image* | Image mapping | object |
+| compoent | Name of the component | string |
```
[the remark this answers is not in the document]
Comment (davidmogar) [reply definitions/push-application-snapshot/README.md:38]: Yes, please. As I commented before is better if we remove the whole `### applicationSnapshot` section.

File: definitions/push-application-snapshot/README.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+| pullspec | Image registry to pull the component image | string |
+| repository | Image registry to push the component image | string |
+| tag | string to tag images image at `repository` | string |
+
+## Remarks:
```
Comment (davidmogar) [inline definitions/push-application-snapshot/README.md:43]:
```suggestion
## Remarks
```

File: definitions/skopeo-copy/skopeo-copy.yaml [hunk 13]
```diff
@@ -39,7 +39,7 @@ spec:
           skopeo copy \
             --all \
             --preserve-digests \
-            --retry-times "$(params.retries)" \
+            --retry-times="$(params.retries)" \
```
Comment (davidmogar) [inline definitions/skopeo-copy/skopeo-copy.yaml:42]: This file should not be in this PR.

**Review by davidmogar — CHANGES_REQUESTED** [review]

File: definitions/push-application-snapshot/README.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
+# push-application-snapshot
+This [Tekton Task](https://tekton.dev/docs/pipelines/tasks/#overview) is responsible for pushing the components defined by the [applicationSnapshot](#applicationSnapshot) from one registry to another.
```
Comment (johnbieren) [inline definitions/push-application-snapshot/README.md:2]: I don't think we need a link to the tekton task documentation here

**Comment by johnbieren** [comment]
@parasense a lot of comments are being resolved here without them being replied to nor addressed. It is making the review pretty hard to follow

**Review by johnbieren — CHANGES_REQUESTED** [review]
- Can you add the README.md back to this PR?
- I think the rename of the `skopeo-copy-task.yaml` file should not be part of this PR. If we want to rename it, it should be a separate PR. I think that is what David was getting at before as well
- There are still some things in `push-application-snapshot.yaml` that are not addressed

**Review by johnbieren — DISMISSED** [review]
Can we get the README.md back in this PR and the skopeo-copy task out of it?

**Comment by parasense** [comment]
The acceptance criteria in [Jira](https://issues.redhat.com/browse/HACBS-564) seems complete.
Also, I believe this material was used in a recent demo, so presumably it's complete.

Here is a copy of the criteria with check boxes added: 
- [x] A push-application-snapshot bundle gets created with a push-application-snapshot task.
- [x] The task expects, at least, one parameter called mappedApplicationSnapshot with the JSON content of the ApplicationSnapshot and the target repository and, optionally, tag where to push the image.
- [x] The task will perform the same procedure seen in skopeo-copy-task for each item in the mappedApplicationSnapshot array.
- [x] The push-images bundle gets renamed to skope-copy.
- [x] The skopeo-copy-task.yaml file gets renamed to skopeo-copy.yaml
- [x] The skopeo-copy-task task gets renamed to skopeo-copy-task.

Re-requesting the review, and ask for somebody to please approve the PR so the Jira issue could be to moved to resolved.
Thanks in advance.

**Comment by davidmogar** [comment]
This PR is not ready for a review. As I said multiple times before, the skopeo renaming should not be part of this PR as it's not related. This PR is for adding a new bundle. Please, remove it and I'll be happy to review it.

Also, I'd like to see the README.md file back as we are trying to make an effort to make simpler for the user and developers to understand what a task does without taking a look to the code. Nevertheless, for the shake of this PR, I'm ok with adding that later on.

**Comment by davidmogar** [comment]
Let's move forward and leave this behind as probably what I expect was not clear enough. We can discuss those details in a future meeting to agree on how contributions should be done. Starting my review now.

**Review by davidmogar — APPROVED** [review]

**Comment by davidmogar** [comment]
Please rebase and merge once the CI passes.

**Comment by davidmogar** [comment]
> Can we get the README.md back in this PR and the skopeo-copy task out of it?

Let's do this in the close future. Now we should focus on the pre-M6 clean up.

**Comment by davidmogar** [comment]
Squash and sign please.
